src/pymor/algorithms/randrangefinder.py

```diff
@@ -0,0 +1,138 @@
+from scipy.linalg import eigh
+from scipy.special import erfinv
+
+from pymor.algorithms.gram_schmidt import gram_schmidt
+from pymor.core.defaults import defaults
+from pymor.core.logger import getLogger
+from pymor.operators.interfaces import OperatorInterface
+from pymor.vectorarrays.numpy import NumpyVectorSpace
+from pymor.operators.numpy import NumpyMatrixOperator
+
+
+@defaults('tol', 'failure_tolerance', 'num_testvecs')
+def adaptive_rrf(A, source_product=None, range_product=None, tol=1e-4, failure_tolerance = 1e-15, num_testvecs = 20, iscomplex=False):
+    """Adaptive randomized range approximation of `A`.
+    
+    This is an implementation of Algorithm 1 in [BS18]_.
+
+    Given the |Operator| `A`, the return value of this method is the |VectorArray|
+    `B` with the property 
+    
+    .. math:: 
+        \Vert A - P_{span(B)} A \Vert \leq tol 
+        
+    with a failure probability smaller than `failure_tolerance`, where the inner product of the range 
+    of `A` is given by `range_product` and the inner product of the source of `A` 
+    is given by `source_product`.
+
+    Parameters
+    ----------
+    A
+        The |Operator| A.
+    source_product
+        Inner product |Operator| of the source of A.
+    range_product
+        Inner product |Operator| of the range of A.
+    tol
+        Error tolerance for the algorithm.
+    failure_tolerance
+        Maximum failure probability.
+    num_testvecs
+        Number of test vectors. 
+    iscomplex
+        If `True`, the random vectors are chosen complex. 
+
+    Returns
+    -------
+    B
+        |VectorArray| which contains the basis, whose span approximates the range of A.
+    """
+    
+    assert source_product is None or isinstance(source_product, OperatorInterface)
+    assert range_product is None or isinstance(range_product, OperatorInterface)
+    
+    logger = getLogger('pymor.algorithms.randrangefinder.adaptive_rrf')
+    
+    B = A.range.empty()
+    M = A.range.empty()
+    
+    for i in range(num_testvecs):
+        r = A.source.random(distribution='normal')
+        if iscomplex:
+            r += 1j*A.source.random(distribution='normal')
+        M.append(A.apply(r))
+    
+    if source_product is None:
+        lambda_min = 1
+    elif issparse(source_product.matrix):
+        lambda_min = eigsh(source_product.matrix, which="SM", return_eigenvectors=False, k=1)[0]
+    else:
+        lambda_min = eigh(source_product.matrix.todense(), eigvals_only=True)[0]
+    
+    testlimit = erfinv( (failure_tolerance / min(A.source.dim, A.range.dim))**(1./num_testvecs)) * tol * np.sqrt(2. * lambda_min)
+    maxnorm = float('inf')
+    
+    while(maxnorm > testlimit):
+        basis_length = len(B)
+        v = A.source.random(distribution='normal')
+        if iscomplex:
+            v += 1j*A.source.random(distribution='normal')
+        B.append(A.apply(v))
+        gram_schmidt(B, range_product, offset=basis_length, copy=False)
```

Comment:
I guess here in `gram_schimdt`, `atol` and `rtol` should be set to zero, or they should be optional parameters of `adaptive_rrf`...